### ISSUE TITLE
Question: how is hypercert artwork generated

### ISSUE THREAD

**Issue by ccerv1** [Network-Goods/hypercerts#69]
Note: this issue is related to Issue #51, but specifically about how and where the hypercert artwork is generated.

We have three options to consider:

***

**Option 1: Static SVG [stored offchain]**

- Artwork can be dynamically generated from user inputs, including logos, icons, etc
- Artwork is stored on IPFS with the rest of the hypercert's metadata
- Artwork is updated in the event of split & merge if performed through our SDK

Pros: 
- Simplicity
- Gives creators more control over artwork and design aspects

Cons:
- There can be counterfeits / scams if splits & merges are performed outside our SDK
- The value in the `fraction` field can get stale if other owners perform splits & merges

***

**Option 2: Dynamic Rich SVG [only practical for L2s]*

- SVG is generated on-chain from the hypercert parameters 
- SVG format is not optimized so it would be very expensive to mint on Mainnet

Pros: 
- Dynamic SVG art will always reflect latest state of the hypercert
- Cool factor without need to be as gas-conscious

Cons:
- Only practical on L2s for the foreseeable future

***

**Option 3: Dynamic Sparse SVG [gas-efficient]**

- SVG is generated on-chain from the hypercert parameters 
- SVG format is highly optimized to be gas-efficient on mainnet (minimal letters, simple vectors, etc)

Pros: 
- Dynamic SVG art will always reflect latest state of the hypercert
- Cool factor

Cons:
- Significant tradeoffs between visual appeal and gas efficiency

**Comment by ccerv1** [comment]
Here's a mockup of the three options (everything is dummy data, including the vector artwork for Option 3)
![SIDE-BY-SIDE](https://user-images.githubusercontent.com/42869436/208777369-5bd13dff-4ccd-442f-9b8e-97499d0143ce.png)

**Comment by bitbeckers** [comment]
Some technical considerations and calculations since the three options presented don’t fully resonate with me:

### Patterns

Current SVG rendering is costly once. Specifically on storing the background patterns which are basically paths described as string. A solution to this is finding a formula/pattern to generate circles. 

Example generator: https://fffuel.co/cccoil/ Generates: 
```xml
<svg xmlns="http://www.w3.org/2000/svg" version="1.1"
    xmlns:xlink="http://www.w3.org/1999/xlink"
    xmlns:svgjs="http://svgjs.dev/svgjs" viewBox="0 0 800 800">
    <defs>
        <linearGradient x1="50%" y1="0%" x2="50%" y2="100%" id="cccoil-grad">
            <stop stop-color="hsl(206, 75%, 49%)" stop-opacity="1" offset="0%"></stop>
            <stop stop-color="hsl(331, 90%, 56%)" stop-opacity="1" offset="100%"></stop>
        </linearGradient>
    </defs>
    <g stroke="url(#cccoil-grad)" fill="none" stroke-linecap="round">
        <circle r="363" cx="400" cy="400" stroke-width="7" stroke-dasharray="1939 2281" transform="rotate(360, 400, 400)" opacity="0.05"></circle>
        <circle r="346.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="1762 2177" transform="rotate(343, 400, 400)" opacity="0.10"></circle>
        <circle r="330" cx="400" cy="400" stroke-width="7" stroke-dasharray="1595 2073" transform="rotate(326, 400, 400)" opacity="0.14"></circle>
        <circle r="313.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="1435 1970" transform="rotate(309, 400, 400)" opacity="0.19"></circle>
        <circle r="297" cx="400" cy="400" stroke-width="7" stroke-dasharray="1284 1866" transform="rotate(291, 400, 400)" opacity="0.23"></circle>
        <circle r="280.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="1141 1762" transform="rotate(274, 400, 400)" opacity="0.28"></circle>
        <circle r="264" cx="400" cy="400" stroke-width="7" stroke-dasharray="1007 1659" transform="rotate(257, 400, 400)" opacity="0.32"></circle>
        <circle r="247.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="881 1555" transform="rotate(240, 400, 400)" opacity="0.37"></circle>
        <circle r="231" cx="400" cy="400" stroke-width="7" stroke-dasharray="764 1451" transform="rotate(223, 400, 400)" opacity="0.41"></circle>
        <circle r="214.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="655 1348" transform="rotate(206, 400, 400)" opacity="0.46"></circle>
        <circle r="198" cx="400" cy="400" stroke-width="7" stroke-dasharray="554 1244" transform="rotate(189, 400, 400)" opacity="0.50"></circle>
        <circle r="181.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="462 1140" transform="rotate(171, 400, 400)" opacity="0.55"></circle>
        <circle r="165" cx="400" cy="400" stroke-width="7" stroke-dasharray="378 1037" transform="rotate(154, 400, 400)" opacity="0.59"></circle>
        <circle r="148.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="302 933" transform="rotate(137, 400, 400)" opacity="0.64"></circle>
        <circle r="132" cx="400" cy="400" stroke-width="7" stroke-dasharray="235 829" transform="rotate(120, 400, 400)" opacity="0.68"></circle>
        <circle r="115.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="176 726" transform="rotate(103, 400, 400)" opacity="0.73"></circle>
        <circle r="99" cx="400" cy="400" stroke-width="7" stroke-dasharray="126 622" transform="rotate(86, 400, 400)" opacity="0.77"></circle>
        <circle r="82.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="84 518" transform="rotate(69, 400, 400)" opacity="0.82"></circle>
        <circle r="66" cx="400" cy="400" stroke-width="7" stroke-dasharray="50 415" transform="rotate(51, 400, 400)" opacity="0.86"></circle>
        <circle r="49.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="25 311" transform="rotate(34, 400, 400)" opacity="0.91"></circle>
        <circle r="33" cx="400" cy="400" stroke-width="7" stroke-dasharray="8 207" transform="rotate(17, 400, 400)" opacity="0.95"></circle>
        <circle r="16.5" cx="400" cy="400" stroke-width="7" stroke-dasharray="0 104" opacity="1.00"></circle>
    </g>
</svg>
```
Notice that the patterns approach becomes pretty clear. Render circles and parameterise certain variables.

### Cost

Rendering content for the SVG can be costly in multiple places:
    *  Name + description; storing strings is simply expensive, but there’s a lot of space between a title or a full abstract. One solution could be to limit all string to bytes32 since we can then store the string as bytes32 which is more gas efficient.
        * “As a general rule, use bytes for arbitrary-length raw byte data and string for arbitrary-length string (UTF-8) data. If you can limit the length to a certain number of bytes, always use one of the value types bytes1 to bytes32 because they are much cheaper.” 
         Ref: https://docs.soliditylang.org/en/develop/types.html#bytes-and-string-as-arrays
        * Doing some testing in Remix ‘this is a pretty long string’ is on the limit between ~25K gas and ~50K gas which at current prices (gas: 11, ETH 1217) between 0.33USD and 0.66USD. After that, testing with a 50 word Lorum Ipsum increases gas cost to ~52K.
        Ref: https://ethereum.stackexchange.com/questions/101975/is-it-cheaper-to-store-a-string-or-convert-it-to-bytes32-with-abi-encodepacked Storing
        * Conclusion: Storing strings smaller than 32 bytes is a low-cost option. Looking at ETH ATH prices (not gas!) you can say ~1USD per string. 
        Ref for prices: https://www.cryptoneur.xyz/gas-fees-calculator
    * Impact labels; if they become part of the SVG, we should also store these on-chain. Besides the ’strings are expensive’ case, you can also imagine issues with localisation (future problem?) and spelling differences (we lightly gatekeep that in the pilot app but lower-case-kebab-casing all input). 
    * General data; we opted to put most data of-chain so that minting would be as cheap as possible. So putting data on-chain again, depending on the number of datapoints, can increase initial minting cost again. But, we also had a heavy protocol and iirc paid high gas-prices for deployment.

Given the points above, the difference between 2 and 3 is not as large as it (visually) implies. Adding custom logo’s, as in option 1, might become expensive on the user side because they also need to be aware of the cost of storing that data on-chain and how to generate their images efficiently. Having some ‘logo-registry’ would make that a one-time cost but it kinda deviates from Ethereal as a state/calculating machine.

### TL;DR:

For option 3 compared to current spec (v0.2) we'd need to add :
- Domain string (Ethereum infrastructure)
- Name string (ETH Fusion)
- 2 uint256 for timestamps
- Smart generation of circle patterns; define and parameterize formula

For option 2 compared to the points above:
- Description string (A Highly... Construction)
- 2 impact labels

**Comment by ccerv1** [comment]
@bitbeckers thanks for this. 

Effectively it comes down to "should the svg be dynamic + generated onchain" ?

And, with the first option (off chain), is it true that:
1) We can only ensure that splits/merges generate the correct new SVG if they are performed through our SDK
2) There is no way to update SVGs of related hypercerts (eg, the fraction displayed on my hypercert SVG might become stale if you do splits/merges on your share of the hypercert)

**Comment by ccerv1** [comment]
Brief update from standup just now:

- It would be weird to have different contract implementations depending on the chain [eg, one version on Mainnet, one version on Polygon]
- It would be less weird to have two contract implementations available on any chain: one that has a static SVG; one that has a dynamic SVG

**Comment by ccerv1** [comment]
OK, just had discussion with @holkeb and @ryscheng to resolve this.

We're going with Option 1:
- SVG will be designed in browser through our dApp
- When the project is ready to create its hypercert, the SVG will be stored on IPFS 
- When users mint their hypercert, they will all look the same (the fractional badge on the bottom right of the image will not be shown)

Rationale:
- Give projects more control over how their hypercerts look
- Make it more realistic to deploy on Mainnet from Day 1
- We can always explore dynamic options in the future


![SIDE-BY-SIDE (2)](https://user-images.githubusercontent.com/42869436/208998122-bff87b7b-5f09-4562-9f0d-b2dede9783c6.png)

**Comment by ccerv1** [comment]
Closing this out
